src/main/java/me/chrr/scribble/mixin/BookEditScreenMixin.java

```diff
@@ -497,31 +597,91 @@ private void setPageContent(String newContent) {
         Scribble.LOGGER.warn("setPageContent() was called, but ignored.");
     }
 
+    @Inject(method = "charTyped", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/util/SelectionManager;insert(Ljava/lang/String;)V"), cancellable = true)
+    private void charTypedEditMode(char chr, int modifiers, CallbackInfoReturnable<Boolean> cir) {
+        Command command = new InsertTextCommand<>(this, currentPageSelectionManager, Character.toString(chr));
+        commandManager.execute(command);
+        cir.setReturnValue(true);
+        cir.cancel();
+    }
+
     @Inject(method = "keyPressedEditMode", at = @At(value = "HEAD"), cancellable = true)
     private void keyPressedEditMode(int keyCode, int scanCode, int modifiers, CallbackInfoReturnable<Boolean> cir) {
-        // Copy/cut/paste without formatting when SHIFT is held down.
-        if (hasControlDown() && hasShiftDown() && !hasAltDown()) {
-            if (keyCode == GLFW.GLFW_KEY_C) {
-                this.getRichSelectionManager().copyWithoutFormatting();
-            } else if (keyCode == GLFW.GLFW_KEY_X) {
-                this.getRichSelectionManager().cutWithoutFormatting();
-            } else if (keyCode == GLFW.GLFW_KEY_V) {
-                this.getRichSelectionManager().pasteWithoutFormatting();
+        // Override copy-without-formatting shortcut
+        if (hasControlDown() && hasShiftDown() && !hasAltDown() && keyCode == GLFW.GLFW_KEY_C) {
+            getRichSelectionManager().copyWithoutFormatting();
+            cir.setReturnValue(true);
+            cir.cancel();
+            return;
+        }
+
+        // Override CUT and CUT-without-formatting shortcut
+        if (hasControlDown() && !hasAltDown() && keyCode == GLFW.GLFW_KEY_X) {
+            Command command = new ActionCommand<>(this, () -> {
+                if (hasShiftDown()) {
+                    getRichSelectionManager().cutWithoutFormatting();
+                } else {
+                    getRichSelectionManager().cut();
+                }
+            });
+            commandManager.execute(command);
+            cir.setReturnValue(true);
+            cir.cancel();
+            return;
+        }
+
+        // Override PASTE and PASTE-without-formatting shortcut
+        if (hasControlDown() && !hasAltDown() && keyCode == GLFW.GLFW_KEY_V) {
+            // ToDo try to find more elegant way how to implement PASTE command
+            //  which won't use internal selectionManager.paste/pasteWithoutFormatting() implementation
+            String textToPaste = hasShiftDown() ? Formatting.strip(getRawClipboard()) : getRawClipboard();
+            Command command = new InsertTextCommand<>(this, getRichSelectionManager(), textToPaste);
+            commandManager.execute(command);
+            cir.setReturnValue(true);
+            cir.cancel();
+            return;
+        }
+
+        // Inject hotkeys for Undo and Redo
```

Comment:
Something I noticed while testing: undos/redos aren't saved properly:

1. Have a book with 2 pages of text
2. Change something on the second page, and immediately undo that.
3. Go to the first page and press Done.

When you open the book again, you'll see that the second page is saved as if it's not undone.

Reply:
Thanks for spotting. I fixed it by [setting](https://github.com/wiskiw/ScribblePlusMC/blob/923c0adac62a207eeb9d8fb1a0abce93bbe1374f/src/main/java/me/chrr/scribble/mixin/BookEditScreenMixin.java#L203) updated(restored) page content to `pages` list in this [commit](https://github.com/wiskiw/ScribblePlusMC/commit/923c0adac62a207eeb9d8fb1a0abce93bbe1374f#diff-29e4ca3dd6093092bb95628d6a75610aadc89f3d219a9fa8787ef44b5b0e8d5eR203). 
Looks like it solved the issue.